spec/lib/page_cloner_spec.rb

```diff
@@ -0,0 +1,111 @@
+require 'rails_helper'
+
+describe PageCloner do
+  let!(:tag)  { create(:tag) }
+  let(:campaign) { create(:campaign) }
+  let(:page)  { create(:page, tags: [tag], campaign: campaign) }
+  let!(:link) { create(:link, page: page) }
+
+  subject(:cloned_page) { PageCloner.clone!(page) }
+
+  it 'clones page' do
+    expect(cloned_page.id).not_to eq(page.id)
+  end
+
+  it 'clones links' do
+    expect(cloned_page.links).not_to eq(page.links)
+    expect(cloned_page.links.count).to eq(1)
+  end
+
+  it 'clones tag associations' do
+    expect(cloned_page.tags).to eq(page.tags)
+    expect(cloned_page.tags.count).to eq(1)
+  end
+
+  it 'associates with the same language' do
+    expect(cloned_page.language).to eq(page.language)
+    expect(cloned_page.language).not_to be_nil
+  end
+
+  it 'associates with the same campaign' do
+    expect(cloned_page.campaign).to eq(campaign)
+  end
+
+  describe 'title and slug' do
+    # TODO: not sure what's best here.
+
+    it 'appends timestamp to title' do
+      expect(cloned_page.title).to match(/\d{4}-\d{1,2}-\d{1,2}/)
+    end
+  end
+
+  context 'images' do
+    let!(:image) { create(:image, page: page) }
+    let!(:primary_image) { create(:image, page: page) }
+
+    before do
+      page.update(primary_image: primary_image)
+    end
+
+    it 'clones image' do
+      expect(cloned_page.images.first).not_to eq(page.images.first)
+    end
+
+    it 'clones primary image' do
+      expect(page.primary_image).to eq(primary_image)
+      expect(cloned_page.primary_image).not_to eq(primary_image)
+    end
```

Comment:
would be good to add a spec that the new and old pages have the same number of images (as I think it would fail that spec currently), probably worth adding that spec for images and plugins etc too